app/models/piece.rb

```diff
@@ -2,75 +2,62 @@
 
 class Piece < ApplicationRecord
   belongs_to :game
+  
+  def is_white?
+    piece_number < 6
+  end
+
+  def is_black?
+    piece_number > 5
+  end
 
   def is_obstructed?(x, y)
     x_sorted_array = [x, x_position].sort
     y_sorted_array = [y, y_position].sort
 
-    #if piece is in current position
-    if (x == self.x_position) && (y == self.y_position) 
-      false
+    obstructions = game.pieces.find do |chess_piece|
+      next true if has_diagonal_obstruction?(x, y, chess_piece, x_sorted_array, y_sorted_array)
+      next true if chess_piece.x_position.between?(x_sorted_array[0] + 1, x_sorted_array[1] - 1)
 
-    #Vertical
-    elsif y == self.y_position 
-      @pieces_in_the_way = []
-      self.game.pieces.each do |chess_piece|
-        if chess_piece != self
-          if (chess_piece.y_position == y) && chess_piece.x_position.between?(x_sorted_array[0], x_sorted_array[1])
-            @pieces_in_the_way << chess_piece
-          end
-        end
-      end
-      !@pieces_in_the_way.empty?
+      chess_piece.y_position.between?(y_sorted_array[0] + 1, y_sorted_array[1] - 1)
 
-    #Horizontal
-    elsif x == self.x_position 
-      @pieces_in_the_way = []
-      self.game.pieces.each do |chess_piece|    if chess_piece != self
-          if (chess_piece.x_position == x) && chess_piece.y_position.between?(y_sorted_array[0], y_sorted_array[1])
-            @pieces_in_the_way << chess_piece
-          end
-        end
-      end
-      !@pieces_in_the_way.empty?
-
-    #Diagonal
-    elsif (x - self.x_position).abs == (y-self.y_position).abs
-      @pieces_in_the_way = self.game.pieces.select do |chess_piece|
-        next false if chess_piece == self
-        ((x - chess_piece.x_position).abs == (y - chess_piece.y_position).abs) && chess_piece.x_position.between?(x_sorted_array[0], x_sorted_array[1]) && chess_piece.y_position.between?(y_sorted_array[0], y_sorted_array[1])
-      end
-      !@pieces_in_the_way.empty?
+    end
 
+    return obstructions.present? || players_own_piece_is_there?(x, y) 
```

Comment:
I like having the players_own_piece_is_there? method here, that way when we begin to render available moves, this position would not show up since the players own colored piece is there.

We might need to move players_own_piece_is_there? to the method that renders available positions in the future, to keep the method definitions clear. It might look something like:

`def render_available_moves(piece)`
`  is_obstructed?`
`  players_own_piece_is_there?`
`  # Logic for determining available moves`
`end`

Reply:
Would you not just call `is_valid?` for that piece? Since the `is_valid?` methods check for if something is `is_obstructed?` and `is_obstructed?` checks if `players_own_piece_is_there?`. The two exceptions would be Knight and King but as we discussed in DM's it would be a simple change to add checks for `players_own_piece_is_there?` to each of those. You have more experience with the rendering of pieces so I could be off base here.

Reply:
There are any number of ways you could organize it, but I think in order to keep the method simple and direct it should not call another method. It makes more sense to have another method call them both at the same time, when that new method is determining where a piece can move and then displaying those options for the player. 

It would also reduce the amount of times we have `is_obstructed?` and `players_own_piece_is_there?` called in the code-base, which supports DRY.